eng/scripts/live-test-resource-cleanup.ps1

```diff
@@ -3,93 +3,373 @@
+        return $true
+      }
+      return $false
+    }
+    if ($ShouldCheckAllowList) {
+      $lines = Get-Content $AllowListPath
+      foreach ($line in $lines) {
+        if ($line -and !$line.StartsWith("#")) {
+          $Exceptions.Add($line.Trim())
+        }
+      }
+      # If allow list is empty or only contains comments, disable allow list checking
+      if (!$Exceptions.Count) {
+        $ShouldCheckAllowList = $false
+        return $false
+      }
+      return HasException $ResourceGroup
+    }
+    return $false
+}
+
+function FindOrCreateDeleteAfterTag {
+  [CmdletBinding(SupportsShouldProcess = $true, ConfirmImpact = 'Medium')]
+  param(
+    [object]$ResourceGroup
+  )
+
+  $deleteAfter = GetTag $ResourceGroup "DeleteAfter"
+  if (!$deleteAfter -or !($deleteAfter -as [datetime])) {
+    $deleteAfter = [datetime]::UtcNow.AddHours($DeleteAfterHours)
+    if ($Force -or $PSCmdlet.ShouldProcess("$($ResourceGroup.ResourceGroupName) [DeleteAfter (UTC): $deleteAfter]", "Update Group DeleteAfter tag")) {
+      Write-Host "Adding DeleteAfer tag with value '$deleteAfter' to group '$($ResourceGroup.ResourceGroupName)'"
+      $ResourceGroup | Update-AzTag -Operation Merge -Tag @{ DeleteAfter = $deleteAfter }
+    }
+  }
+}
+
+function HasDeleteLock() {
+  return $false
+}
+
+function DeleteOrUpdateResourceGroups() {
+  [CmdletBinding(SupportsShouldProcess = $true, ConfirmImpact = 'Medium')]
+  param()
+
+  if ($IsProvisionerApp) {
+    AddGithubUsersToAliasCache
+  }
+
+  Write-Verbose "Fetching groups"
+  $allGroups = @(Get-AzResourceGroup)
+  $toDelete = @()
+  $toUpdate = @()
+  Write-Host "Total Resource Groups: $($allGroups.Count)"
+
+  foreach ($rg in $allGroups) {
+    if (HasException $rg) {
+      continue
+    }
+    $deleteAfter = GetDeleteAfterTag $rg
+    if ($deleteAfter) {
+      if (HasExpiredDeleteAfterTag $deleteAfter) {
+        $toDelete += $rg
+      }
+      continue
+    }
+    if (!$DeleteNonCompliantGroups) {
+      continue
+    }
+    if (HasValidAliasInName $rg) {
+      continue
+    }
+    if (HasValidOwnerTag $rg -or HasDeleteLock $rg) {
+      continue
+    }
+    $toUpdate += $rg
+  }
+
+  foreach ($rg in $toUpdate) {
+    FindOrCreateDeleteAfterTag $rg
+  }
+
+  # Get purgeable resources already in a deleted state.
+  $purgeableResources = @(Get-PurgeableResources)
+
+  Write-Host "Total Resource Groups To Delete: $($toDelete.Count)"
+  foreach ($rg in $toDelete)
+  {
+    $deleteAfter = GetTag $rg "DeleteAfter"
+    if ($Force -or $PSCmdlet.ShouldProcess("$($rg.ResourceGroupName) [DeleteAfter (UTC): $deleteAfter]", "Delete Group")) {
+      # Add purgeable resources that will be deleted with the resource group to the collection.
+      $purgeableResourcesFromRG = Get-PurgeableGroupResources $rg.ResourceGroupName
+
+      if ($purgeableResourcesFromRG) {
+        $purgeableResources += $purgeableResourcesFromRG
```

Comment:
I think this already existed but won't purgeable resources block the RG from being deleted? So won't that step fail? Do we need to try to purge these before attempting to delete the RG that contains them? Or are these resources outside of the RG?

Reply:
@heaths Do you know?

Reply:
At least for purgeable KV resources, once purged they are no longer part of any RG.

Reply:
I think what @weshaggard is saying is, do we need to do the purge before trying to do the RG delete. Or is the resource group delete what puts the resource into the purgeable state?

Reply:
Deleting an RG with KVs puts them into a purgeable state. But the two functions here are for different things. `Get-PurgeableGroupResources` is getting those KVs that *will* be in a purged state once the RG is deleted. The other function gets all the already-deleted KVs. Since it takes time to actually delete a KV and put it in a purgeable state, my only options were to do what I did with 2 different functions, or put some sort of wait/sleep in the script with an arbitrary time that still may not work. That's not deterministic, hence 2 functions.

Reply:
Ahh I see. That makes sense thanks for clarifying.

Also @weshaggard I think from experience, a purgeable resource won't prevent the group from deleting. If I recall correctly we've had scenarios where someone deleted a group with keyvaults in them, and in order to restore the soft-deleted keyvault, the deleted group was also restored on the backend.

Reply:
Yes, you are correct there as well: you can delete an RG which will delete a KV and put it in a purgeable state. That will not affect the...deletedness?...of RGs.